File: github_code/quick-sort/cpp/6.c
/*****************************************************************************
Taylor Hoss
X432Z869
Program #10

This is a program that implements four sorting algoritms and a binary search
tree. It will generate data sets of 100, 500, and 1000 via arrays using a
datatype that holds a key to be sorted by and a value that acts as a
placeholder for data. As each sort goes through each data set it tells how
many iterations it went through to get to the sorted list. Also when
the data is added to the binary search tree and later transversed in-order
it gives iteration counts as well. All of these iteration counts can be
compared to the Big-O worst-case calculations.

*****************************************************************************/

#ifndef IOSTREAM
#define IOSTREAM

#include <iostream>

#endif //IOSTREAM

#ifndef SORT_DATA_CPP
#define SORT_DATA_CPP

#include "sort_data.cpp"

#endif //SORT_DATA_CPP

#ifndef FSTREAM
#define FSTREAM

#include <fstream>

#endif //FSTREAM

#ifndef STRING
#define STRING

#include <string>

#endif //STRING

#ifndef NAMESPACE_STD
#define NAMESPACE_STD

using namespace std;

#endif //NAMESPACE_STD

//function that implements the quick sort algorithm
int quickSort(sort_data sortAry[], int left, int right, int iter)
{
   int i = left, j = right, mid = ( ( left + right ) / 2 );
   sort_data swap;
   
   //finding pivot
   if (sortAry[left].key > sortAry[mid].key)
   {
      swap = sortAry[mid];
      sortAry[mid] = sortAry[left];
      sortAry[left] = swap;
   }

   if (sortAry[left].key > sortAry[right].key)
   {
      swap = sortAry[right];
      sortAry[right] = sortAry[left];
      sortAry[left] = swap;
   }

   if (sortAry[mid].key > sortAry[right].key)
   {
      swap = sortAry[right];
      sortAry[right] = sortAry[mid];
      sortAry[mid] = swap;
   }

   int pivot = sortAry[mid].key;

   //partioning set
   while (i <= j)
   {
      while (sortAry[i].key < pivot)
      {
         i++;
      }
      while (sortAry[j].key > pivot)
      {
         j--;
      }
    
      if (i <= j)
      {
         swap = sortAry[i];
         sortAry[i] = sortAry[j];
         sortAry[j] = swap;

         i++;
         j--;
      }

      iter++;
   }

   //recursively calling for partitions of partitions
   if (left < j)
      iter = quickSort(sortAry, left, j, iter);
   if (i < right)
      iter = quickSort(sortAry, i, right, iter);

   return iter;
}

File: github_code_index/java/4/6.java
package Sorting;

public class InsertionSort_JAVA {

    public static void main(String[] args) {
        int test[] = {8, 5, 6, 2, 4};
        System.out.println("삽입정렬 전");
        for (int data : test) {
            System.out.print(data + ", ");
        }
        
        insertionSort(test, test.length);
        
        System.out.println("\n삽입정렬 후");
        for (int data : test) {
            System.out.print(data + ", ");
        }
    }
    
    public static void insertionSort(int[] list, int n) {
        int i, j, key;
        
        // 인덱스 0은 이미 정렬된 것으로 볼 수 있다.
        for (i = 1; i < n; i++) {
            key = list[i]; // 현재 삽입될 숫자인 i번쨰 정수를 key 변수로 복사
            
            // 현재 정렬된 배열은 i-1까지이므로 i-1번째부터 역순으로 조사한다.
            // j값은 음수가 아니어야 하고
            // key 값보다 정렬된 배열에 있는 값이 크면 j번째를 j+1번째로 이동
            for (j = i - 1; j >= 0 && list[j] > key; j--) {
                list[j+1] = list[j];
            }
            
            list[j+1] = key;
        }
    }
}
